gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/ManifestBasedDataset.java

```diff
@@ -0,0 +1,145 @@
+          + "%s, you can specify multi locations split by '',", manifestPath.toString(), fs.getUri().toString(), ManifestBasedDatasetFinder.MANIFEST_LOCATION));
+      return Iterators.emptyIterator();
+    }
+    JsonReader reader = null;
+    List<CopyEntity> copyEntities = Lists.newArrayList();
+    List<FileStatus> toDelete = Lists.newArrayList();
+    //todo: put permission preserve logic here?
+    try {
+      reader = new JsonReader(new InputStreamReader(fs.open(manifestPath), "UTF-8"));
+      reader.beginArray();
+      while (reader.hasNext()) {
+        //todo: We can use fileSet to partition the data in case of some softbound issue
+        //todo: After partition, change this to directly return iterator so that we can save time if we meet resources limitation
+        JsonObject file = GSON.fromJson(reader, JsonObject.class);
+        Path fileToCopy = new Path(file.get("fileName").getAsString());
+        if (this.fs.exists(fileToCopy)) {
+          if (!targetFs.exists(fileToCopy) || needCopy(this.fs.getFileStatus(fileToCopy), targetFs.getFileStatus(fileToCopy))) {
+            CopyableFile copyableFile =
+                CopyableFile.fromOriginAndDestination(this.fs, this.fs.getFileStatus(fileToCopy), fileToCopy, configuration)
+                    .fileSet(datasetURN())
+                    .datasetOutputPath(fileToCopy.toString())
+                    .ancestorsOwnerAndPermission(CopyableFile
+                        .resolveReplicatedOwnerAndPermissionsRecursively(this.fs, fileToCopy.getParent(),
+                            new Path("/"), configuration))
+                    .build();
+            copyableFile.setFsDatasets(this.fs, targetFs);
+            copyEntities.add(copyableFile);
+          }
+        } else if (targetFs.exists(fileToCopy)){
+          toDelete.add(targetFs.getFileStatus(fileToCopy));
+        }
+      }
+      if (Boolean.parseBoolean(this.properties.getProperty(DELETE_FILE_NOT_EXIST_ON_SOURCE, "false"))) {
+        CommitStep step = new DeleteFileCommitStep(targetFs, toDelete, this.properties,
+            this.deleteEmptyDirectories ? Optional.of(new Path(this.properties.getProperty(DELETE_EMPTY_DIRECTORIES_UPTO))) : Optional.<Path>absent());
+        copyEntities.add(new PrePublishStep(datasetURN(), Maps.newHashMap(), step, 1));
+      }
+    } catch (JsonIOException| JsonSyntaxException e) {
+      //todo: update error message to point to a sample json file instead of schema which is hard to understand
+      log.warn(String.format("Failed to read Manifest path %s on filesystem %s, please make sure it's in correct json format with schema"
+          + " {type:array, items:{type: object, properties:{id:{type:String}, fileName:{type:String}, fileGroup:{type:String}, fileSizeInBytes: {type:Long}}}}",
+          manifestPath.toString(), fs.getUri().toString()), e);
```

Comment:
missing a `%s` for the exception e

Reply:
log(String, Exception) is the function I'm trying to call